addons/mail/models/mail_mail.py

```diff
@@ -442,23 +459,36 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
             })
 
-        # prepare attachments: remove attachments if user send the link with the
-        # access_token.
         attachments = self.attachment_ids
-        if attachments:
-            if body:
-                link_ids = {int(link) for link in re.findall(r'/web/(?:content|image)/([0-9]+)', body)}
-                if link_ids:
-                    attachments = attachments - self.env['ir.attachment'].browse(list(link_ids))
-            # load attachment binary data with a separate read(), as prefetching all
-            # `datas` (binary field) could bloat the browse cache, triggering
-            # soft/hard mem limits with temporary data.
-            email_attachments = [
-                (a['name'], base64.b64decode(a['datas']), a['mimetype'])
-                for a in attachments.sudo().read(['name', 'datas', 'mimetype']) if a['datas'] is not False
-            ]
-        else:
-            email_attachments = []
+        # Prepare attachments:
+        # Remove attachments if user send the link with the access_token.
+        if body and attachments:
+            link_ids = {int(link) for link in re.findall(r'/web/(?:content|image)/([0-9]+)', body)}
+            if link_ids:
+                attachments = attachments - self.env['ir.attachment'].browse(list(link_ids))
+        # Turn remaining attachments into links if they are too heavy and
+        # their ownership are business models (i.e. something != mail.message,
+        # otherwise they will be deleted along with the mail message leading to a 404)
+        if record_owned_attachments := attachments.sudo().filtered(
+                lambda a: a.res_model and a.res_id and a.res_model != 'mail.message'):
+            estimated_email_size_bytes = self._estimated_email_size(
+                headers, body, [a.file_size for a in attachments.sudo()])
+            max_email_size_bytes = (mail_server or self.env['ir.mail_server']
+                                    ).sudo()._get_max_email_size() * 1024 * 1024
+            if estimated_email_size_bytes > max_email_size_bytes:
+                # Remove attachments and prepare downloadable links to be added in the body
+                record_owned_attachments.sudo().generate_access_token()
+                attachments_links = self.env['ir.qweb']._render('mail.mail_attachment_links',
+                                                                {'attachments': record_owned_attachments})
+                body = tools.append_content_to_html(body, attachments_links, plaintext=False)
+                attachments = self.env['ir.attachment'] - record_owned_attachments
+        # Prepare the remaining attachment (those not embedded as link)
+        # load attachment binary data with a separate read(), as prefetching all
+        # `datas` (binary field) could bloat the browse cache, triggering
+        # soft/hard mem limits with temporary data.
+        email_attachments = [(a['name'], a['raw'], a['mimetype'])
```

Comment:
Hello @awa-odoo , 

I have analysed in details that part and I think you're right. Here is the details:

In any case, we need to loop over the attachments to check if we can/must embed them as query link (to check their global size and check that they are owned by a business record to avoid them to be deleted randomly). So we can’t remove that. Moreover, when embedding we call generate_access_token which also loop over the attachment reading some field.

But following my tests below, the historical “.read” on attachment doesn’t seem to be useful. We could just loop over the record. But following my observation, it doesn’t seems change the number of queries when just looping over the records or reading some specific fields.

Here are my observations:
- When placing a break point in ir_attachment._compute_raw which read the binary data from disk, it is first called when reading explicitly the "raw" with ".read".
- When profiling query for the ".read" statement, no query is generated (when using file system storage, otherwise a query to fetch all the db_datas in one go).
- Replacing the read by a simple loop over the attachments, no additional query is generated as well. And the _compute_raw is also called from there and not before.
- On the contrary, attachments are read when filtering them above as you mention but the content of the file are not loaded (as it is a compute that need an access to it to be triggered and db_datas is a binary field with preftech=False).
- To generate the embedded links, we need to call generate_access_token which loop over attachment while reading some field (similarly to what filter does but potentially on a subset of records).

As far as I understand, binary data can be stored in multiple location, the default one being the file system. But there is also a column "db_datas" that can hold binary data.
If I configure odoo to use it. It doesn't change the behavior: when not reading explicitly the content of the files, it is not read in the cache. This is because binary field have the "prefetch" attribute to False by default.
So it seems that unless the field "raw" (or "db_datas") is accessed explicitly, the file content won't be read. So it seems that using read instead of looping on the attachment records is not useful (and could maybe in some case generate an additional query). Moreover, when debugging the ".read", the db_datas are inserted in the cache.

In conclusion, using read instead of looping over the records seems unnecessary but I might missing something. But keeping that historical statement doesn’t seems to hurt either. 

Should I replace the read by a simple loop on records ? What do you think ?

Reply:
Wow, thank you for the detailed analysis 😄 

I have traced back the initial issue to (10 years ago 👴):
https://github.com/odoo/odoo/commit/61d212298773d0c49048fde9f0a1af9f1a2d1df1

Looks like it's more a problem of "don't try to load the attachments of all emails being sent at once". .
So I would say let's not mess with it and keep it as is.